README.md
# Memcached::RateLimit ![static](https://github.com/uperl/Memcached-RateLimit/workflows/static/badge.svg) ![linux](https://github.com/uperl/Memcached-RateLimit/workflows/linux/badge.svg)

Sliding window rate limiting with Memcached

# SYNOPSIS

```perl
use Memcached::RateLimit;

my $rl = Memcached::RateLimit->new("memcache://localhost:11211");
$rl->error_handler(sub ($rl, $message) {
  warn "rate limit error: $message";
});

# allow 30 requests per minute
if($rl->rate_limit("resource", 1, 30, 60))
{
  # rate limit exceeded
}
```

# DESCRIPTION

This module implements rate limiting logic.  It is intended for high
volume websites that require limits on the access or modification to
resources.  It is implemented using Rust and [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus), so you
will need the rust toolchain in order to install this module.

Why Rust?  Well none of the Perl Memcache clients I found supported
TLS, and the Rust [memcache crate](https://crates.io/crates/memcache)
did.  Also Rust is fast and has a number of safety checks that give
me confidence that it won't crash our app.

The actual algorithm is based one used by Bugzilla, and by default
it will "fail open", meaning if for some reason the client cannot
connect to the Memcached server, it will **allow** the request.

# CONSTRUCTOR

## new

```perl
my $rl = Memcached::RateLimit->new($url);
my $rl = Memcached::RateLimit->new(\%config);
```

Create a new instance of [Memcached::RateLimit](https://metacpan.org/pod/Memcached::RateLimit).  The URL should be of the
form shown in the synopsis above.

The following schemes are supported:

- `memcache`
- `memcache+tcp`
- `memcache+tls`
- `memcache+udp`
- `memcache+unix`

You can append these query parameters
to the URL:

- `connect_timeout`

    Connect timeout in seconds.  May be specified as a
    floating point, that is `0.2` is 20 milliseconds.

- `protocol`

    If set to `ascii` this will use the ASCII protocol instead of binary.

- `tcp_nodelay`

    Boolean `true` or `false`.

- `timeout`

    IO timeout in seconds. May be specified as a
    floating point, that is `0.2` is 20 milliseconds.

- `verify_mode`

    For TLS, this can be set to `none` or `peer`.

\[version 0.03\]

You can provide a `%Config` hash instead of a URL.  All of the
query parameters mentioned above can be provided in addition to
these:

- `scheme`

    The scheme (example: `memcache` or `memcache+tls`).

- `host`

    The server hostname or IPv4/IPv6 address.

- `port`

    The TCP or UDP port to connect to.

- `read_timeout`

    The read timeout in seconds.  May be specified as a
    floating point, that is `0.2` is 20 milliseconds.

- `write_timeout`

    The write timeout in seconds.  May be specified as a
    floating point, that is `0.2` is 20 milliseconds.

- `retry`

    \[version 0.04\]

    The default instance number of retries.

# METHODS

## rate\_limit

```perl
my $limited = $rl->rate_limt($name, $size, $rate_max, $rate_seconds);
my $limited = $rl->rate_limt($name, $size, $rate_max, $rate_seconds, $retry);
```

This method returns a boolean true, if a request of `$size` exceeds the
rate limit of `$rate_max` over the past `$rate_seconds`.  If you only
want to rate limit the number of requests then you can set `$size` to 1.

This method will return a boolean false, and increment the appropriate
counters if the requests fits within the rate limit.

This method will **also** return boolean false, if it is unable to connect
to or otherwise experiences an error talking to the memcached server.
In this case it will also call the [error handler](#error_handler).

\[version 0.04\]

If `$retry` is provided then if there are errors talking to memcached, it
will be attempted `$retry` times.  If this parameter is not provided, then
the default instance retry limit will be used, and if there is not instance
default the class default of `1` will be used.

## set\_read\_timeout

```
$rl->set_read_timeout($secs);
```

Sets the IO Read timeout to `$secs`, may be fractional.

## set\_write\_timeout

```
$rl->set_write_timeout($secs);
```

Sets the IO Write timeout to `$secs`, may be fractional.

## error\_handler

```perl
$rl->error_handler(sub ($rl, $message) {
  ...
});
```

This method will set the error handler, to be called in the case of an
error with the memcached server.  It will pass in the instance of
[Memcached::RateLimit](https://metacpan.org/pod/Memcached::RateLimit) as `$rl` and a diagnostic as `$message`.
Since this module will fail open, it is probably useful to increment
error counters and provide diagnostics with this method to your monitoring
system.

## final\_error\_handler

\[version 0.04\]

```perl
$rl->final_error_handler(sub ($rl, $message) {
});
```

This method is like the [error\_handler method](#error_handler), but it
only gets called at the end if none of the retry attempts succeed.
The last error message is passed in.

# SEE ALSO

- [Cache::Memcached::Fast](https://metacpan.org/pod/Cache::Memcached::Fast)
- [Redis::RateLimit](https://metacpan.org/pod/Redis::RateLimit)

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Dylan Hardison (DHARDISON)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.09      2023-06-14 17:18:02 -0600
  - Use `*std::os::raw::c_char` instead of `*i8` for C strings.
    This fixes the build on Linux Arm64, and possibly other
    platforms (gh#18)

0.08      2023-04-11 12:15:27 -0600
  - Remove ffi/Cargo.lock from git and not distribute, which
    is recommended for rust crate libraries (gh#15, gh#17, Dylan++)

0.07      2022-11-03 06:20:11 -0600
  - Fixed probably a bug on 32 bit CPUs (gh#14)
  - Added checking for NULL string pointers for improved
    safety (gh#14)

0.06      2022-10-31 12:26:26 -0600
  - connect_timeout and fractional timeouts are no longer
    experimental (gh#2, gh#13)

0.05      2022-10-30 06:26:50 -0600
  - Fixed bug where sum could be miscalculated over the
    rate_seconds period (gh#12)
  - Breaking change: the keys used internally have altered
    slightly to make it easier to debug

0.04      2022-10-25 19:40:17 -0600
  - Optional retries added (gh#11)

0.03      2022-10-25 13:06:39 -0600
  - Constructor can now take a config hash reference
    instead of a URL (gh#10)

0.02      2022-10-25 09:05:27 -0600
  - Experimental support for connection timeout,
    and fractional timeouts (gh#1, gh#9)

0.01      2022-10-25 07:32:27 -0600
  - initial version


ffi/src/lib.rs
use anyhow::bail;
use anyhow::Result;
use memcache::Client;
use memcache::CommandError::KeyExists;
use memcache::MemcacheError::CommandError;
use memcache::Url;
use std::cell::Cell;
use std::cell::RefCell;
use std::collections::HashMap;
use std::ffi::CStr;
use std::ffi::CString;
use std::str;
use std::time::Duration;
use std::time::SystemTime;

struct Rl {
    url: Url,
    client: Option<Client>,
    error: CString,
    read_timeout: Option<Duration>,
    write_timeout: Option<Duration>,
}

impl Rl {
    fn new(url: &CStr) -> Result<Rl> {
        let url = Url::parse(url.to_str()?)?;
        Ok(Rl {
            url: url,
            client: None,
            error: CString::new("")?,
            read_timeout: None,
            write_timeout: None,
        })
    }

    fn rate_limit(
        &mut self,
        prefix: &CStr,
        size: u32,
        rate_max: u32,
        rate_seconds: u32,
    ) -> Result<bool> {
        let prefix = prefix.to_str()?;

        /* re-connect if necessary */
        if let None = self.client {
            let client = Client::connect(String::from(self.url.clone()))?;
            client.set_read_timeout(self.read_timeout)?;
            client.set_write_timeout(self.write_timeout)?;
            self.client = Some(client);
        }

        if let Some(client) = &self.client {
            let now = SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)?
                .as_secs();

            let key = format!("{}{}", prefix, now);
            match client.add(&key, 0, rate_seconds + 1) {
                Ok(()) => (),
                Err(CommandError(e)) if e == KeyExists => (),
                Err(e) => bail!(e),
            };

            let keys: Vec<String> = (0..rate_seconds)
                .map(|n| format!("{}{}", prefix, n))
                .collect();
            let keys: Vec<&str> = keys.iter().map(|s| &**s).collect();
            let tokens: HashMap<String, u32> = client.gets(&keys)?;
            let (value, cas_id) = match client.get(&key)? {
                Some((value, _, cas)) => (
                    match str::from_utf8(&value) {
                        Ok(value) => match value.parse::<u32>() {
                            Ok(value) => value,
                            Err(_) => 0,
                        },
                        Err(_) => 0,
                    },
                    cas,
                ),
                None => (0, None),
            };

            let mut sum = value + size;
            for (_, v) in tokens {
                sum += v;
            }

            if sum > rate_max {
                return Ok(true);
            }

            if let Some(cas_id) = cas_id {
                client.cas(&key, value + size, rate_seconds + 1, cas_id)?;
            }

            Ok(false)
        } else {
            // TODO
            // shouldn't get in here?
            // but should probably error?
            Ok(false)
        }
    }
}

thread_local!(
  static COUNTER: Cell<u64> = Cell::new(1);
  static STORE: RefCell<HashMap<u64, Rl>> = RefCell::new(HashMap::new())
);

#[no_mangle]
pub extern "C" fn rl_new(url: *const i8) -> u64 {
    let url = unsafe { CStr::from_ptr(url) };
    let rl = match Rl::new(url) {
        Ok(rl) => rl,
        Err(_) => return 0,
    };

    let index = COUNTER.with(|it| {
        let index = it.get();
        it.set(index + 1);
        index
    });

    STORE.with(|it| {
        let mut it = it.borrow_mut();
        it.insert(index, rl);
    });

    index
}

#[no_mangle]
pub extern "C" fn rl__rate_limit(
    index: u64,
    prefix: *const i8,
    size: u32,
    rate_max: u32,
    rate_seconds: u32,
) -> i32 {
    let prefix = unsafe { CStr::from_ptr(prefix) };

    STORE.with(|it| {
        return match it.borrow_mut().get_mut(&index) {
            Some(rl) => match rl.rate_limit(prefix, size, rate_max, rate_seconds) {
                Ok(true) => 1,
                Ok(false) => 0,
                Err(e) => {
                    if let Ok(error) = CString::new(e.to_string()) {
                        rl.error = error;
                    }
                    rl.client = None;
                    -1
                }
            },
            None => -1,
        };
    })
}

const INVALID_OBJECT_INDEX_ERROR_MESSAGE: *const u8 = b"Invalid object index\0" as *const u8;

#[no_mangle]
pub extern "C" fn rl__error(index: u64) -> *const i8 {
    STORE.with(|it| {
        let error = match it.borrow().get(&index) {
            Some(rl) => rl.error.as_ptr(),
            None => INVALID_OBJECT_INDEX_ERROR_MESSAGE as *const i8,
        };

        error
    })
}

#[no_mangle]
pub extern "C" fn rl_set_write_timeout(index: u64, timeout: f64) {
    STORE.with(|it| {
        if let Some(rl) = it.borrow_mut().get_mut(&index) {
            rl.write_timeout = Some(Duration::from_secs_f64(timeout));
            if let Some(client) = &rl.client {
                let _ = client.set_write_timeout(rl.write_timeout);
            }
        }
    });
}

#[no_mangle]
pub extern "C" fn rl_set_read_timeout(index: u64, timeout: f64) {
    STORE.with(|it| {
        if let Some(rl) = it.borrow_mut().get_mut(&index) {
            rl.read_timeout = Some(Duration::from_secs_f64(timeout));
            if let Some(client) = &rl.client {
                let _ = client.set_read_timeout(rl.read_timeout);
            }
        }
    });
}

#[no_mangle]
pub extern "C" fn rl_DESTROY(index: u64) {
    STORE.with(|it| {
        it.borrow_mut().remove(&index);
    })
}


lib/Memcached/RateLimit.pm
use warnings;
use 5.020;
use experimental qw( postderef signatures );

package Memcached::RateLimit {

  # ABSTRACT: Sliding window rate limiting with Memcached

  use FFI::Platypus 2.00;
  use Ref::Util qw( is_plain_hashref );
  use Carp qw( croak );

  my $ffi = FFI::Platypus->new( api => 2, lang => 'Rust' );
  $ffi->bundle;
  $ffi->mangler(sub ($name) { "rl_$name" });
  $ffi->type("object(@{[ __PACKAGE__ ]},u64)" => 'rl');
  our %retry;
  our %error_handler;
  our %final_error_handler;

  sub _hash_to_url (%config)
  {
    my %q;
    my $scheme          = delete $config{scheme}            // 'memcache';
    my $host            = delete $config{host}              // '127.0.0.1';
    my $port            = delete $config{port}              // '11211';
    my $read_timeout    = delete $config{read_timeout};
    my $write_timeout   = delete $config{write_timeout};
    my $retry           = delete $config{retry};
    $q{connect_timeout} = delete $config{connect_timeout}   if defined $config{connect_timeout};
    $q{protocol}        = delete $config{protocol}          if defined $config{protocol};
    $q{tcp_nodelay}     = delete $config{tcp_nodelay}       if defined $config{tcp_nodelay};
    $q{timeout}         = delete $config{timeout}           if defined $config{timeout};
    $q{verify_mode}     = delete $config{verify_mode}       if defined $config{verify_mode};

    require URI::Escape;

    croak("Unknown options: @{[ sort keys %config ]}") if %config;
    # host may need to be escaped if it is a IPv6 address
    my $url = "$scheme://@{[ URI::Escape::uri_escape($host) ]}:$port";

    if(%q)
    {
      # In theory none of the query parameters should have characters that need to be
      # escaped, but since we have to pull in uri_escape for the hostname, we may as
      # well escape these too.
      $url .= "?" . join '&', map { join '=', $_, URI::Escape::uri_escape($q{$_}) } sort keys %q;
    }

    ($url, $read_timeout, $write_timeout, $retry);
  }

  $ffi->attach( new => ['string'] => 'u64' => sub ($xsub, $class, $url) {

    my $read_timeout;
    my $write_timeout;
    my $retry;

    ($url, $read_timeout, $write_timeout, $retry) = _hash_to_url(%$url)
      if is_plain_hashref $url;

    my $index = $xsub->($url);
    my $self = bless \$index, $class;

    $retry{$$self} = $retry if defined $retry;

    $self->set_read_timeout($read_timeout) if defined $read_timeout;
    $self->set_write_timeout($write_timeout) if defined $write_timeout;

    $self;
  });

  $ffi->attach( _rate_limit       => ['rl','string','u32','u32','u32'] => 'i32' );
  $ffi->attach( _error            => ['rl'] => 'string'                         );
  $ffi->attach( set_read_timeout  => ['rl', 'f64']                              );
  $ffi->attach( set_write_timeout => ['rl', 'f64']                              );

  $ffi->attach( DESTROY => ['rl'] => sub ($xsub, $self) {
    delete $error_handler{$$self};
    delete $final_error_handler{$$self};
    delete $retry{$$self};
  });

  sub error_handler ($self, $sub)
  {
    $error_handler{$$self} = $sub;
  }

  sub final_error_handler ($self, $sub)
  {
    $final_error_handler{$$self} = $sub;
  }

  sub rate_limit ($self, $name, $size, $rate_max, $rate_seconds, $retry=undef)
  {
    $retry //= $retry{$$self} // 1;
    my $error;
    for(1..$retry)
    {
      my $ret = _rate_limit($self, $name, $size, $rate_max, $rate_seconds);
      if($ret == -1)
      {
        $error_handler{$$self}->($self, $error = $self->_error) if defined $error_handler{$$self};
        next;
      }
      else
      {
        return $ret;
      }
    }
    # fail open
    $final_error_handler{$$self}->($self, $error //= $self->_error) if defined $final_error_handler{$$self};
    return 0;
  }

}

1;

=head1 SYNOPSIS

 use Memcached::RateLimit;
 
 my $rl = Memcached::RateLimit->new("memcache://localhost:11211");
 $rl->error_handler(sub ($rl, $message) {
   warn "rate limit error: $message";
 });
 
 # allow 30 requests per minute
 if($rl->rate_limit("resource", 1, 30, 60))
 {
   # rate limit exceeded
 }

=head1 DESCRIPTION

This module implements rate limiting logic.  It is intended for high
volume websites that require limits on the access or modification to
resources.  It is implemented using Rust and L<FFI::Platypus>, so you
will need the rust toolchain in order to install this module.

Why Rust?  Well none of the Perl Memcache clients I found supported
TLS, and the Rust L<memcache crate|https://crates.io/crates/memcache>
did.  Also Rust is fast and has a number of safety checks that give
me confidence that it won't crash our app.

The actual algorithm is based one used by Bugzilla, and by default
it will "fail open", meaning if for some reason the client cannot
connect to the Memcached server, it will B<allow> the request.

=head1 CONSTRUCTOR

=head2 new

 my $rl = Memcached::RateLimit->new($url);
 my $rl = Memcached::RateLimit->new(\%config);

Create a new instance of L<Memcached::RateLimit>.  The URL should be of the
form shown in the synopsis above.

The following schemes are supported:

=over 4

=item C<memcache>

=item C<memcache+tcp>

=item C<memcache+tls>

=item C<memcache+udp>

=item C<memcache+unix>

=back

You can append these query parameters
to the URL:

=over 4

=item C<connect_timeout>

Connect timeout in seconds.  May be specified as a
floating point, that is C<0.2> is 20 milliseconds.

=item C<protocol>

If set to C<ascii> this will use the ASCII protocol instead of binary.

=item C<tcp_nodelay>

Boolean C<true> or C<false>.

=item C<timeout>

IO timeout in seconds. May be specified as a
floating point, that is C<0.2> is 20 milliseconds.

=item C<verify_mode>

For TLS, this can be set to C<none> or C<peer>.

=back

[version 0.03]

You can provide a C<%Config> hash instead of a URL.  All of the
query parameters mentioned above can be provided in addition to
these:

=over 4

=item C<scheme>

The scheme (example: C<memcache> or C<memcache+tls>).

=item C<host>

The server hostname or IPv4/IPv6 address.

=item C<port>

The TCP or UDP port to connect to.

=item C<read_timeout>

The read timeout in seconds.  May be specified as a
floating point, that is C<0.2> is 20 milliseconds.

=item C<write_timeout>

The write timeout in seconds.  May be specified as a
floating point, that is C<0.2> is 20 milliseconds.

=item C<retry>

[version 0.04]

The default instance number of retries.

=back

=head1 METHODS

=head2 rate_limit

 my $limited = $rl->rate_limt($name, $size, $rate_max, $rate_seconds);
 my $limited = $rl->rate_limt($name, $size, $rate_max, $rate_seconds, $retry);

This method returns a boolean true, if a request of C<$size> exceeds the
rate limit of C<$rate_max> over the past C<$rate_seconds>.  If you only
want to rate limit the number of requests then you can set C<$size> to 1.

This method will return a boolean false, and increment the appropriate
counters if the requests fits within the rate limit.

This method will B<also> return boolean false, if it is unable to connect
to or otherwise experiences an error talking to the memcached server.
In this case it will also call the L<error handler|/error_handler>.

[version 0.04]

If C<$retry> is provided then if there are errors talking to memcached, it
will be attempted C<$retry> times.  If this parameter is not provided, then
the default instance retry limit will be used, and if there is not instance
default the class default of C<1> will be used.

=head2 set_read_timeout

 $rl->set_read_timeout($secs);

Sets the IO Read timeout to C<$secs>, may be fractional.

=head2 set_write_timeout

 $rl->set_write_timeout($secs);

Sets the IO Write timeout to C<$secs>, may be fractional.

=head2 error_handler

 $rl->error_handler(sub ($rl, $message) {
   ...
 });

This method will set the error handler, to be called in the case of an
error with the memcached server.  It will pass in the instance of
L<Memcached::RateLimit> as C<$rl> and a diagnostic as C<$message>.
Since this module will fail open, it is probably useful to increment
error counters and provide diagnostics with this method to your monitoring
system.

=head2 final_error_handler

[version 0.04]

 $rl->final_error_handler(sub ($rl, $message) {
 });

This method is like the L<error_handler method|/error_handler>, but it
only gets called at the end if none of the retry attempts succeed.
The last error message is passed in.

=head1 SEE ALSO

=over 4

=item L<Cache::Memcached::Fast>

=item L<Redis::RateLimit>

=back

=cut
